### PR TITLE
docker/build.sh; update flags, allow GCR images.

### PR DESCRIPTION
**Pull request by eap**
Make it easier for users of Google Container Registry to build
images. Also make the command line parsing a little less sensitive
to order and less dependent on environmental vars for config.

### CODE REVIEW

**Comment by rtitle** [comment]
Will do!

**Comment by helgridly** [comment]
@jacmrob can you take a look at this too?

File: docker/build.sh [hunk 1]
```diff
@@ -9,24 +9,132 @@
...
+                echo "note: separate args from flags with a space, not '='."
+            fi
+            exit 1
+            ;;
+    esac
+    shift
+done
+
+# Print help after all flags are parsed successfully
+if PRINT_HELP; then
```
Comment (rtitle) [inline docker/build.sh]: Typo, `if $PRINT_HELP; then` ?

**Comment by rtitle** [comment]
Looks good. I successfully pushed images to Dockerhub and GCR from my laptop. Thanks for leaving the default behavior unchanged. 

Recently for https://github.com/DataBiosphere/leonardo/issues/142 I started some work to allow the Dataproc init script to pull the Jupyter image from GCR instead of Dockerhub. So this is great timing. I can try to make a PR for those changes, and then we can officially support GCR!

File: docker/build.sh (hunk 1 above)
Comment (eap) [reply docker/build.sh]: Oh good catch; messed with that right before pushing - probably should have re-tested.

**Comment by rtitle** [comment]
Actually I plugged this into our automated build/test infrastructure (looks like it doesn't automatically kick off for PRs from forks) and it failed with:

```
building jupyter docker image...
invalid argument "leonardo/leonardo-notebooks:origin/rt-update-build-scripts-for-gcr" for t: Error parsing reference: "leonardo/leonardo-notebooks:origin/rt-update-build-scripts-for-gcr" is not a valid repository/tag
```
https://fc-jenkins.dsp-techops.broadinstitute.org/job/leonardo-build/1117/console

I'll look into why..

File: docker/build.sh [hunk 2]
```diff
@@ -9,24 +9,132 @@
...
+\t
+EOF
+)"
+
+# Enable strict evaluation semantics.
+set -e
+
+# Set default variables used while parsing command line options.
+TARGET="${TARGET:-leonardo}"
+GIT_BRANCH="${GIT_BRANCH:-$(git rev-parse --abbrev-ref HEAD)}"
```
Comment (rtitle) [inline docker/build.sh]: Our Jenkins infrastructure sets `$BRANCH` not `$GIT_BRANCH`, so I needed to change this to the following:
```
GIT_BRANCH="${BRANCH:-$(git rev-parse --abbrev-ref HEAD)}"
```

File: docker/build.sh [hunk 3]
```diff
@@ -9,24 +9,132 @@
...
+
+# Print help after all flags are parsed successfully
+if $PRINT_HELP; then
+  echo -e "${HELP_TEXT}"
+  exit 0
+fi
+
+# Configure script using arguments.
+if [[ $DOCKER_REGISTRY == "dockerhub" ]]; then
+  PROJECT="${PROJECT:-broadinstitute}"
```
Comment (rtitle) [inline docker/build.sh]: Heh, our Jenkins also sets `PROJECT=leonardo`. So `IMAGE` comes out as `leonardo/leonardo`. :)

Could we maybe call it `DOCKER_PROJECT` or something instead?

**Comment by eap** [comment]
I don't have permission to merge, if this looks good to y'all, feel free to do the squash and merge.

**Pull request merged**